### PR TITLE
Bluetooth: controller: Add reset of per adv for HCI_Reset command

### PR DESCRIPTION
**Pull request by ppryga-nordic**
There is missing handling of periodic advertising reset for
HCI_Reset command. That makes impossible to execute e.g.
qualification tests without rebooting of the DUT
for periodic advertising and direction finding.

This commit adds missing implementation.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_adv.c [hunk 1]
```diff
@@ -1506,6 +1507,13 @@ int ull_adv_reset_finalize(void)
 	uint8_t handle;
 	int err;
 
+#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
+	err = ull_adv_sync_reset_finalize();
+	if (err) {
+		return err;
+	}
+#endif /* CONFIG_BT_CTLR_ADV_PERIODIC */
+
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_adv.c:1516]: `_finalise()` is not required, as lll_adv_data_reset for periodic advertising is called on every `ll_adv_sync_param_set` call.

```suggestion
```

File: subsys/bluetooth/controller/ll_sw/ull_adv_internal.h [hunk 2]
```diff
@@ -150,6 +150,7 @@ void ull_adv_aux_ptr_fill(uint8_t **dptr, uint8_t phy_s);
 
 int ull_adv_sync_init(void);
 int ull_adv_sync_reset(void);
+int ull_adv_sync_reset_finalize(void);
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_adv_internal.h:153]:
```suggestion
```

File: subsys/bluetooth/controller/ll_sw/ull_adv_sync.c [hunk 3]
```diff
@@ -635,8 +635,62 @@ int ull_adv_sync_init(void)
...
+			sync->is_enabled = 0U;
+
+			continue;
+		}
+
+		err = sync_remove(sync, adv, 0U);
+		if (err) {
+			return err;
+		}
+	}
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:667]:
```suggestion
		}
		
		adv->lll.sync = NULL;
	}
```

File: subsys/bluetooth/controller/ll_sw/ull_adv_sync.c [hunk 4]
```diff
@@ -635,8 +635,62 @@ int ull_adv_sync_init(void)
...
+		}
+
+		lll_sync = adv->lll.sync;
+		if (!lll_sync) {
+			continue;
+		}
+
+		lll_adv_data_reset(&lll_sync->data);
+		adv->lll.sync = NULL;
+	}
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:692]:
```suggestion
```

File: subsys/bluetooth/controller/ll_sw/ull_df.c [hunk 5]
```diff
@@ -124,6 +124,25 @@ int ull_df_reset(void)
...
+		adv = ull_adv_is_created_get(handle);
+		if (!adv) {
+			continue;
+		}
+
+		if (adv->df_cfg) {
+			adv->df_cfg->is_enabled = 0U;
+			ull_df_adv_cfg_release(adv->df_cfg);
+			adv->df_cfg = NULL;
+		}
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_df.c:142]: Set to `NULL` is sufficient as `lll_df_adv_cfg_pool` is `mem_init` in `init_reset` function.

```suggestion
		adv->df_cfg = NULL;
```

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_adv.c (hunk 1 above)
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c:1516]: It seems to me that it is required to clear `adv->lll.sync`.

File: subsys/bluetooth/controller/ll_sw/ull_adv_sync.c (hunk 4 above)
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:692]: Referring to comment: https://github.com/zephyrproject-rtos/zephyr/pull/37333#discussion_r680801307

I suggest to do not remove the function. Just change it to set all adv->lll.sync to NULL.
So this part may be removed:
```
lll_sync = adv->lll.sync;
if (!lll_sync) {
	continue;
}
lll_adv_data_reset(&lll_sync->data);
```

**Comment by ppryga** [comment]
@cvinayak I've refactored code to follow sequence of reset:
- reset ULL context
- reset LLL context
- reset and initialize LLL static variables, performed by `perform_lll_reset`
- reset ULL static variables,
- init ULL context variables.

File: subsys/bluetooth/controller/ll_sw/ull_df.c (hunk 5 above)
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_df.c]: Unnecessary code removed

File: subsys/bluetooth/controller/ll_sw/ull_adv_sync.c (hunk 3 above)
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:667]: `adv->lll.sync = NULL` moved to `ull_adv_reset_finalize` following pattern with `adv->lll.aux=NULL`.

File: subsys/bluetooth/controller/ll_sw/ull_adv_internal.h (hunk 2 above)
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_adv_internal.h:153]: Function not removed.

File: subsys/bluetooth/controller/ll_sw/ull_adv.c (hunk 1 above)
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c]: The code is left here.

**Comment by ppryga** [comment]
@cvinayak take a look on the PR please.

File: subsys/bluetooth/controller/ll_sw/ull_adv.c [hunk 6]
```diff
@@ -1506,14 +1495,34 @@ int ull_adv_reset_finalize(void)
...
+	}
+#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
+	err = ull_adv_sync_reset_finalize();
+	if (err) {
+		return err;
+	}
+#endif /* CONFIG_BT_CTLR_ADV_PERIODIC */
+#endif /* CONFIG_BT_CTLR_ADV_AUX_SET */
+#endif /* CONFIG_BT_CTLR_ADV_EXT */
+
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_adv.c:1514]:
```suggestion
#if defined(CONFIG_BT_CTLR_ADV_EXT)
#if defined(CONFIG_BT_CTLR_ADV_AUX_SET)
	if (CONFIG_BT_CTLR_ADV_AUX_SET > 0) {
		err = ull_adv_aux_reset_finalize();
		if (err) {
			return err;
		}
#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
		err = ull_adv_sync_reset_finalize();
		if (err) {
			return err;
		}
#endif /* CONFIG_BT_CTLR_ADV_PERIODIC */
	}
#endif /* CONFIG_BT_CTLR_ADV_AUX_SET */
#endif /* CONFIG_BT_CTLR_ADV_EXT */

```

File: subsys/bluetooth/controller/ll_sw/ull_adv.c [hunk 7]
```diff
@@ -1506,14 +1495,34 @@ int ull_adv_reset_finalize(void)
...
 		struct lll_adv *lll = &adv->lll;
 
 #if defined(CONFIG_BT_CTLR_ADV_EXT)
 		adv->is_created = 0;
 		lll->aux = NULL;
-#endif
+#endif /* CONFIG_BT_CTLR_ADV_EXT */
+#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
+		lll->sync = NULL;
+#endif /* CONFIG_BT_CTLR_ADV_PERIODIC */
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_adv.c:1525]:
```suggestion
#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
		lll->sync = NULL;
#endif /* CONFIG_BT_CTLR_ADV_PERIODIC */
#endif /* CONFIG_BT_CTLR_ADV_EXT */
```

**Review by cvinayak — CHANGES_REQUESTED** [review]
Minor conditional compilation changes

File: subsys/bluetooth/controller/ll_sw/ull_adv.c (hunk 6 above)
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c]: Change applied.

File: subsys/bluetooth/controller/ll_sw/ull_adv.c (hunk 7 above)
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c]: Change applied.

**Comment by ppryga** [comment]
@cvinayak PR updated.

**Review by cvinayak — APPROVED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_adv_aux.c [hunk 8]
```diff
@@ -434,7 +434,7 @@ int ull_adv_aux_init(void)
 	return 0;
 }
 
-int ull_adv_aux_reset(void)
+int ull_adv_aux_reset_finalize(void)
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:437]: I am not sure I understand this renaming: Without a `ull_adv_aux_reset`, or a `ull_adv_aux_reset_start` or anything like that, what purpose does the qualifier `finalize` have? 

I see that we have a `ull_adv_reset_finalize` but we also have a `ull_adv_reset` which doesn't seem to be the case for  `ull_adv_aux`

**Review by Thalley — APPROVED** [review]
A question regarding naming, otherwise LGTM

File: subsys/bluetooth/controller/ll_sw/ull_adv_aux.c (hunk 8 above)
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:437]: `ull_adv_reset` is responsible for stopping all scheduling in ULL and request to stop scheduling LLL.
After that myflay is used to call LLL to reset all statics and globals.
Then `ull_adv_reset_finalize` enters to reset and initialize ULL globals.

I didn't change the naming conventions, just change other functions that actually do ULL memory clear and initialization.

**Pull request merged**